File: BootCamp Java/Gun/Gun.java
public class Gun {
    private int amount;
    private int capacity;
    private boolean isReady;
    private String model;
    private int totalShots;
    
    public Gun (String model, int capacity) {
        this.amount = 0;
        this.capacity = capacity;
        this.isReady = false;
        this.model = model;
        this.totalShots = 0;
    }
    public Gun() {
        this("Beretta", 8);
    }
    public int getAmount() {
        return this.amount;
}

    public int getCapacity() {
        return this.capacity;
    }

    public boolean ready() {
        return this.isReady;
    }

    public String getModel() {
        return this.model;
    }

    public int getTotalShots() {
        return this.totalShots;
    }

    public void prepare() {
        this.isReady = !isReady;
    }

    public void reload() {
        this.amount = this.capacity;
    }

    public void shoot() throws NotReadyException, OutOfRoundsException {
        if (!isReady) {
            throw new NotReadyException ();
        }
        if ( this.amount == 0 ) {
            throw new OutOfRoundsException();
        }
        System.out.println ("Bang!");
        this.amount -= 1;
        this.totalShots += 1;

    }

public String toString() {
    return "Gun model:" + getModel()+'\n'+"Capacity:"+getCapacity()+'\n'+"Total shots:"+getTotalShots()+'\n'+"Amount:"+getAmount()+'\n'+"Ready:"+ready()+'\n';
    }

}

File: BootCampOOP/Gun/Gun.cpp
#include "Gun.h"

Gun::Gun(const std::string& modelIn, int capacityIn){
    model = modelIn;
    capacity = capacityIn;
    amount = 0;
    totalShots = 0;
    isReady = false;
}

Gun::~Gun(){
  
}

int Gun::getAmount() const{
    return amount;
}

int Gun::getCapacity() const{
    return capacity;
}

bool Gun::ready() const{
    return isReady;
}

const std::string& Gun::getModel() const{
    return model;
}

int Gun::getTotalShots() const{
    return totalShots;
}

void Gun::prepare() {
    isReady = !isReady;
}

void Gun::reload() {
    amount = 0;
}

void Gun::shoot() {
    if ( !ready() ) {
       throw NotReady();
    }
    if ( amount == 0 ) {
       throw OutOfRounds();
    }
    std::cout<< "BANG!" << std::endl;
    std::cout<< "" << std::endl;
    amount -= 1;
    totalShots += 1; 
    isReady = false; 
}

std::ostream& operator<<(std::ostream& out, const Gun& gun){
    out << "Модель:" << gun.getModel() << std::endl;
    out << "Кол-во встрелов:" << gun.getTotalShots() << std::endl;
    out << "Вместимость магазина:" << gun.getCapacity() << std::endl;
    out << "Кол-во патронов:" << gun.getAmount() << std::endl;
}